### PR TITLE
Incorporated FE Workflow into Archetype

### PR DESCRIPTION
**Pull request by jbmatheson**
### Features
Full Typescript, ES6, ES5 and jQuery Support (with applicable Webpack wrappers)
TS and JS linting (following TSLint ruleset – to be refined)
ES5 output (for legacy browser support)
Globbing:
- No need to add imports anywhere
- All JS and CSS files are added to each component (best practice is under /clientlib/js or clientlib/(s)css
- No .content.xml or js.txt/css.txt files needed as everything is run through Webpack
- The globber pulls in all JS files under the component folder. Webpack allows CSS/SCSS files to be chained in via JS files. So, They are pulled in through the two entry points, sites.js and vendors.js.
- The only file consumed by AEM is the output files site.js and site.css in /clientlibsite as well as dependencies.js and dependencies.css in /clientlib-dependencies
- Chunking
- Main (site js/css)
- Vendors (dependencies js/css)
Output:
- Site.js/Site.css à clientlib-site folder
- Dependencies.js/Dependencies.css à clientlib-dependencies folder
JS Optimization - When run through the production mode all JS that is not being used or
called is removed.
CSS:
- Full Sass/Scss support
- Sass is compiled to CSS
- Autoprefixing – All CSS is run through a prefixer and any properties that require prefixing will automatically have those added in the CSS.
- Minified (both dev and prod)
- CSS Optimization (via cssnano) – At post, all CSS is run through an optimizer, which normalizes all CSS based on the following rules:
    - Reduces CSS calc expression wherever possible, ensuring both browser compatibility and compression
    - Converts between equivalent length, time & angle values. Note that by default, length values are not converted.
    - Removes comments in and around rules, selectors & declarations.
    - Removes duplicated rules, at-rules and declarations. Note that this only works for exact duplicates.
    - Removes empty rules, media queries & rules with empty selectors, as they do not affect the output.
    - Merges adjacent rules by selectors & overlapping property/value pairs.
    - Ensures that only a single @charset is present in the CSS file and moves it to the top of the document. This prevents multiple, invalid declarations occurring through naïve CSS concatenation.
    - Replaces the CSS initial keyword with the actual value, when the resulting output is smaller
Compresses inline SVG definitions with SVGO.
- Explicit clean task for wiping out the generated CSS, JS and Map files on demand
- Source Mapping (development mode only)
- Separate NPM script to compile only Sass

### Installation
- Install Node (v10+). This will also install NPM.
- At the root of the project run: `npm install`
### Usage
There are four NPM scripts that drive the FE workflow:

`npm run dev`
`npm run prod`
`npm run tsc`
`npm run sass`
### Scripts
`npm run dev` - Runs the full build with:
- no JS optimization (tree shaking, etc)
- Source maps are available
- CSS minified

`npm run prod` - Runs the full build with:
- JS optimization enabled
- Source maps disabled
- CSS minified

`npm run tsc` - Runs only the TS/JS linter to identify errrors/warnings (note: the ruleset is still being
updated).
- Requires you to have TSLint installed globally. Remember to use the global flag when
installing: npm i tslint -g

`npm run sass` - Runs only the Sass compilation (same configuration as the main build.)
- Utilizes a dev-only webpack config file.
- Same configuration, otherwise, as the main build

### CODE REVIEW

**Comment by jbmatheson** [comment]
CLA form has been fiiled out but the check still seems to be failing when I reopen the PR.

**Comment by bpauli** [comment]
@adobe/open-source-office Can you please check why the CLA check is still failing? @jbmatheson reopened the PR after signing the form.
Thanks!

**Comment by stevengill** [comment]
Hey @bpauli

Looks like we are having some issues with Adobe Sign Search API. I can confirm @jbmatheson has signed the CLA correctly. We have shared the issue with the Adobe sign team so they can figure out a solution.

File: src/main/resources/META-INF/maven/archetype-metadata.xml [hunk 1]
```diff
@@ -250,5 +250,17 @@
         <include>.gitignore</include>
       </includes>
     </fileSet>
+    <fileSet filtered="true" encoding="UTF-8">
```
Comment (vladbailescu) [inline src/main/resources/META-INF/maven/archetype-metadata.xml]: Wouldn't these fit better in https://github.com/adobe/aem-project-archetype/blob/d1086465b65d06144424a3499595c52ca333732a/src/main/resources/META-INF/maven/archetype-metadata.xml#L122-L161 ?

**Comment by vladbailescu** [comment]
@jbmatheson, thanks for the valuable contribution! Could you please also include usage instructions for the front-end build? Perhaps add a new section to the project readme that explains how to activate the build, switch between dev and prod and different CSS/JS options.

**Comment by jbmatheson** [comment]
@vladbailescu thank you for the suggestion!  I can add the maven profiles applicable to the Frontend build to the readme. The rest of the usage should be list above in the pr. If that information is enough then I can move it to the readme file, too.

File: src/main/resources/META-INF/maven/archetype-metadata.xml (hunk 1 above)
Comment (jbmatheson) [reply src/main/resources/META-INF/maven/archetype-metadata.xml]: Fixed and updated this to have the fileset includes under `ui.apps`

@vladbailescu

**Comment by jbmatheson** [comment]
@vladbailescu - Archetype README has been update to reflect all FE features and build information.

**Comment by godanny86** [comment]
hi @jbmatheson great work!

@jbmatheson , @vladbailescu I propose introducing a new maven module to the archetype vs. having the fed workflow inside the ui.apps module. I posted my thoughts/reasons in detail on issue #124 for greater visibility. Curious to get your thoughts!

**Comment by bpauli** [comment]
@jbmatheson Thank you very much for this contribution! I'm looking forward to merge it. Do you think you can work on the valuable feedback from @godanny86 in the next days? I think it would be also good to have an additional parameter in the archetype generation to optionally disable/remove the new frontend module and to just have a "plain AEM" project. WDYT?

Of course if you don't have capacity in the next days I'm happy to merge your PR as is and to do the changes afterwards.

**Pull request merged**